### PR TITLE
Fix Appa, Steadfast Guardian

### LINKED ISSUE #8760: [Card Bug] Appa, Steadfast Guardian: airbend any number is parsed as one required target
## Card(s)

Appa, Steadfast Guardian

## Build/version

Not provided.

## Game mode

Unknown

## Actual behavior

When Appa enters, the airbend prompt requires exactly one target and accepts no more than one. The controller cannot choose zero targets, and cannot exile multiple eligible permanents.

## Expected behavior

Appa should allow any number of other target nonland permanents the controller controls, including zero and every eligible permanent.

## Steps to reproduce

1. Put Appa, Steadfast Guardian onto the battlefield while controlling two or more other nonland permanents.
2. Resolve its enters-the-battlefield trigger.
3. Attempt to submit zero targets, then attempt to select more than one target.
4. Observe that the prompt requires one target and does not allow a second.

## Oracle text

> Flash
> Flying
> When Appa enters, airbend any number of other target nonland permanents you control.

## Triage analysis

Confirmed by code inspection as a parser target-cardinality defect. The specialized airbend parser calls `strip_optional_target_prefix`, which recognizes only the `up to N` form. It does not recognize `any number of`. The shared post-parse cardinality extractors do not recover this case because `airbend` is not in `MULTI_TARGET_VERBS`.

Consequently the generated ability has no `multi_target` specification. The zone-change resolution path defaults a missing specification to one required target, matching both reported symptoms.

## Acceptance criteria

- Appa parses with `multi_target: MultiTargetSpec::unlimited(0)`.
- Its trigger can resolve with zero targets.
- With multiple legal permanents, it can select and exile multiple targets.
- Target filter remains other nonland permanents you control.
- Add a parser regression test and a cast/trigger-resolution test covering zero and multiple selected targets.

## Related

- #4384 and #3322 cover distinct airbend behavior already fixed.
- #7610 is an open, distinct target-identity bug for Monk Gyatso.

### PR DESCRIPTION
**Pull request by ludwikgorski-lab**
## Summary

Fix issue #8760: `strip_optional_target_prefix` now recognizes `any number of [other|another] target …` as `MultiTargetSpec::unlimited(0)`, so Appa, Steadfast Guardian's ETB airbend (and other capturing callers of that combinator) no longer default to one required target.

Closes #8760

## Files changed

- `crates/engine/src/parser/oracle_effect/lower.rs`
- `crates/engine/src/parser/oracle_effect/tests.rs`
- `crates/engine/tests/integration/issue_8760_airbend_any_number.rs`
- `crates/engine/tests/integration/main.rs`

## Track

Developer

## LLM

Model: grok-4.6
Tier: Frontier
Thinking: high

## Implementation method (required)

Method: /engine-implementer

> [!NOTE]
> Any change to `crates/engine/` game logic — parser, effects, resolver,
> targeting, rules behavior — is expected to go through `/engine-implementer`.
> The "not used" box is for changes that genuinely fall outside that scope.

## CR references

- CR 107.1c (authorizing: choose "any number" may be any positive number or zero)
- CR 115.6 (a targeted spell or ability may allow zero targets)
- CR 115.1d + CR 603.3d (triggered targeting: targets chosen as the ability is put on the stack)
- CR 115.1a + CR 601.2c (spell optional targeting, existing function docs)
- CR 701.65a (airbend keyword action)

## Verification

- [x] Required checks ran clean, or the exact CI-owned alternative is stated below.
- [x] Gate A output below is for the current committed head.
- [x] Final review-impl below is clean for the current committed head.
- [x] Both anchors cite existing analogous code at the same seam.

- `cargo fmt --all` — clean (ran before commit)
- `cargo clippy -p phase-engine --all-targets -- -D warnings` — PASS (`Finished dev profile in 19m 42s`)
- `cargo test -p phase-engine --lib strip_optional_target_prefix` — 7 passed
- `cargo test -p phase-engine --lib parse_airbend` — 5 passed
- `cargo test -p phase-engine --test integration issue_8760` — 3 passed
- `cargo coverage` — `total_cards=35849 supported_cards=31939 coverage_pct=89.093` (Appa was already `supported: true`; this is a cardinality misparse, not a coverage flip)
- `./scripts/check-parser-combinators.sh aa6948a4f7970661e700bb653f218b542763b32d` — Gate A PASS at head `6b4ff270173afbbda23d6f13162251778df0830b`

## Gate A

Gate A PASS head=6b4ff270173afbbda23d6f13162251778df0830b base=aa6948a4f7970661e700bb653f218b542763b32d

## Anchored on

- crates/engine/src/parser/oracle_effect/lower.rs:7631 — existing `strip_leading_quantifier` maps `any number of ` → `MultiTargetSpec::unlimited(0)` via nom `tag`
- crates/engine/src/parser/oracle_effect/mod.rs:5464 — existing `try_parse_airbend_clause` already captures `strip_optional_target_prefix`'s spec onto `clause.multi_target`

## Final review-impl

Final review-impl PASS head=6b4ff270173afbbda23d6f13162251778df0830b

## Claimed parse impact

Appa, Steadfast Guardian

## Scope Expansion

None.

## Validation Failures

None.

## CI Failures

None.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved parsing for “any number of target,” “other target,” and “another target” effects.
  * Correctly distinguishes target-selection phrases from resource-count phrases.
  * Airbend effects now trigger only when at least one object is exiled.
  * Supports zero-target and multi-target airbend resolution.
  * Improved targeting for nonland permanents and “another” target filters.

* **Tests**
  * Added coverage for optional selections, chained abilities, and airbend-trigger behavior.

### CODE REVIEW

**Review by matthewevans — CHANGES_REQUESTED** [review]
## Changes requested — regression tests need to prove the production selection path

Reviewed at current head `6b4ff270173afbbda23d6f13162251778df0830b`.

1. **The zero-target case can pass without exercising Appa's ETB.** `appa_etb_airbend_zero_selected` calls `runner.cast(appa).resolve()` and then only observes priority and unchanged zones (`crates/engine/tests/integration/issue_8760_airbend_any_number.rs:164-187`). The generic scenario driver accepts an empty set whenever `min_targets == 0` (`crates/engine/src/game/scenario.rs:3631-3658`), so the same assertions pass if the ETB is skipped before it ever reaches target selection. Please drive the real cast/trigger path to `WaitingFor::MultiTargetSelection`, assert that prompt (including its zero minimum), submit the empty selection, then retain the final-state assertions.

2. **The positive case does not prove the legal-target boundary.** `appa_etb_airbend_multiple_selected` declares only the two valid Bears (`issue_8760_airbend_any_number.rs:191-209`), but `SpellCast::target_objects` records intent and the scenario driver filters that intent against `legal_targets` before submitting it (`crates/engine/src/game/scenario.rs:3640-3657`). A filter that incorrectly also admits the land, the opponent's permanent, or Appa still exiles the two Bears and passes. At the production `MultiTargetSelection` boundary, assert that the legal set includes both Bears and excludes the land, opponent permanent, and Appa before selecting and resolving.

The parser helper itself is at the existing shared seam and the exact-head parse receipt is present; this request is solely for discriminating pipeline coverage. CodeRabbit independently identified the same two gaps in its current-head review.

File: crates/engine/src/game/effects/register_bending.rs [hunk 1]
```diff
@@ -13,11 +17,42 @@ pub fn resolve(
...
+fn bend_performed(state: &GameState, ability: &ResolvedAbility, kind: BendingType) -> bool {
+    match kind {
+        BendingType::Air => {
+            let exiled_this_way = QuantityExpr::Ref {
+                qty: QuantityRef::FilteredTrackedSetSize {
+                    filter: Box::new(TargetFilter::Any),
+                    caused_by: Some(ThisWayCause::Exiled),
+                },
+            };
+            resolve_quantity_with_targets(state, &exiled_this_way, ability) > 0
```
[the remark this answers is not in the document]
Comment (matthewevans) [reply crates/engine/src/game/effects/register_bending.rs:57]: Confirmed in code, but no printed card can reach it, and it isn't a regression. `publish_tracked_set` extends the chain's tracked set, and that set resets only at chain entry, CR 700.2 mode boundaries and repeat iterations. So an earlier exile in the same chain would satisfy a later zero-exile airbend. Before this PR, `register_bending` recorded every airbend unconditionally. The new gate can only remove Airbend events, never add one. Census of the 14 airbend cards in card-data: none has two airbends, or an exile ahead of an airbend, in one ability chain. Modal cards reset at the mode boundary, and Whirlwind Technique's earlier discard is excluded by the Exiled-cause gate. A correct fix needs a per-instruction exile scope at the publication seam, and the sibling {2} cast permission (`TrackedSet(0)`) needs the same scope; that predates this PR. Both are tracked together as a maintainer follow-up. At 09023f6f the `bend_performed` doc states the chain scope.

File: crates/engine/src/types/ability.rs [hunk 2]
```diff
@@ -23517,8 +23517,11 @@ pub struct AbilityDefinition {
     /// any-opponent permission. Requires `optional: true`; prompts use APNAP order.
     pub optional_for: Option<OpponentMayScope>,
     /// Variable-count targeting: min/max targets the player can choose.
-    /// When present, resolution enters MultiTargetSelection instead of immediate resolve.
-    /// CR 601.2c + CR 115.1d.
+    /// CR 601.2c + CR 115.1d: when present, target choice on the stack emits one
+    /// `TargetSelectionSlot` per allowed target (up to the resolved max), with
+    /// slots at or above the resolved min marked optional, and surfaces them via
+    /// `WaitingFor::TargetSelection` (spells and activated abilities) or
+    /// `WaitingFor::TriggerTargetSelection` (triggered abilities).
```
[the remark this answers is not in the document]
Comment (matthewevans) [reply crates/engine/src/types/ability.rs]: Fixed at 09023f6f. The doc now cites CR 601.2c (spells) and CR 602.2b (activated) for `TargetSelection`, CR 603.3d (triggered) for `TriggerTargetSelection`, and CR 115.6 for optional slots at or above the minimum. The subject-arm comment now cites CR 115.1 rather than the triggered-only CR 115.1d.

**Review by matthewevans — APPROVED** [review]
Approved at head `09023f6f74fe27417b2588188e84717a4f17912d`.

The parser fix recognizes "any number of [other|another] target" as an unlimited slot with a minimum of zero (CR 107.1c + CR 115.6). Two maintainer fixups sit on top.

**`f919cf6c` counts an airbend only when it exiles something.** CR 701.65b says a player airbends only when they exile one or more objects as a result of the instruction. With the parser fix alone, Appa's zero-target airbend still recorded an airbend, so abilities such as Avatar Aang's draw would trigger on it. `RegisterBending` now checks the chain's "exiled this way" set before recording the airbend. Earthbend keeps counting unconditionally, and firebend and waterbend keep their own paths. The subject-side "any number of" arm now delegates to `strip_optional_target_prefix`, the existing authority for the quantifier, so "another target" composes too. The tests drive the cast pipeline:
- a zero-target Appa airbend (no Aang draw);
- a one-exile airbend (one draw);
- the Avatar Aang transform test, now driven by a real exiling airbend;
- a parser test for "any number of another target".

**`09023f6f` is comments only.**
- The `multi_target` doc cites the rule for each selection path: CR 601.2c for spells, CR 602.2b for activated abilities and CR 603.3d for triggered abilities, plus CR 115.6 for optional slots.
- The subject arm cites CR 115.1 instead of the triggered-only 115.1d.

The 13:07Z changes-requested review (at `6b4ff270`) is addressed at this head. `appa_etb_airbend_zero_selected` and `appa_etb_airbend_multiple_selected` now cast Appa through the real pipeline and stop at its ETB `TriggerTargetSelection` prompt. The zero case asserts that every slot is optional before it submits the empty selection. The positive case asserts that the legal set holds both Bears and excludes the land, the opponent's permanent and Appa, and it does so before selecting.

CodeRabbit's two threads are answered and resolved. The CR-citation finding is fixed. The other finding is valid: the exiled-this-way set is scoped to the resolution chain, not to one instruction. It is not a regression, because the new check can only remove `Airbend` events, never add them. It is also unreachable today: across the 14 printed airbend cards, none has an exile or a second airbend ahead of an airbend in the same resolution. The same over-reach already exists in the older `{2}` cast-permission grant, so both will be scoped per instruction in a follow-up.

Required checks are green at this head.

Thanks for tracking down Appa's optional airbend, @ludwikgorski-lab.

**Review by matthewevans — APPROVED** [review]
Approved at current head `09023f6f74fe27417b2588188e84717a4f17912d`.

The earlier requested changes are resolved: the Appa regression tests now reach the real triggered target-selection prompt, assert zero-minimum slots and the legal-target boundary, then submit both empty and positive selections through the production pipeline. The SHA-bound parse receipt covers the eight affected cards, and current required CI/security checks are green. The later CodeRabbit airbend-chain concern was correctly identified as pre-existing and withdrawn; this change only prevents false Airbend events when nothing was exiled.
